test/hotspot/jtreg/runtime/valhalla/inlinetypes/LarvalMarkWordTest.java

```diff
@@ -0,0 +1,112 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @summary ensure larval bit is kept by Serial
+ * @requires vm.flagless
+ * @enablePreview
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -XX:+UseSerialGC -Xmx50M -Xlog:gc*=info runtime.valhalla.inlinetypes.LarvalMarkWordTest
+ */
+/**
+ * @test
+ * @summary ensure larval bit is kept by Parallel
+ * @requires vm.flagless
+ * @enablePreview
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -XX:+UseParallelGC -Xmx50M -Xlog:gc*=info runtime.valhalla.inlinetypes.LarvalMarkWordTest
+ */
+/**
+ * @test
+ * @summary ensure larval bit is kept by G1
+ * @requires vm.flagless
+ * @enablePreview
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -XX:+UseG1GC -Xmx50M -Xlog:gc*=info runtime.valhalla.inlinetypes.LarvalMarkWordTest
+ */
+/**
+ * @test
+ * @summary ensure larval bit is kept by Z
+ * @requires vm.flagless
+ * @enablePreview
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -XX:+UseZGC -Xmx50M -Xlog:gc*=info runtime.valhalla.inlinetypes.LarvalMarkWordTest
+ */
+
+package runtime.valhalla.inlinetypes;
+
+import jdk.internal.misc.Unsafe;
+
+public class LarvalMarkWordTest {
+  private static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+  // We want to ensure that the larval bit in the markWord is preserved
+  // accross GC events.
+  public static void main(String[] args) throws ReflectiveOperationException {
+    // Take a value class with one field and record its field offset.
+    MyValue val = new MyValue(0);
+    long offset = UNSAFE.objectFieldOffset(val.getClass().getDeclaredField("x"));
+    // Create a new larval value object.
+    val = UNSAFE.makePrivateBuffer(val);
+    // Write a new value to it.
+    UNSAFE.putLong(val, offset, 19);
+    // Now we make a bunch of objects.
+    // Ideally, we want the object to be relocated, but this isn't guaranteed.
+    stressTheGC();
+    // Hopefully the GC has already decided to do a GC, but if not we
+    // suggest it do one. This suggestion may be ignored.
+    System.gc();
+
+    // At this point, the GC's preservation should have seen that val was
+    // a larval, and if it did anything with val, preserved the markWord.
+    // There is a VM assertion that will fail here if the bit is lost.
+    val = UNSAFE.finishPrivateBuffer(val);
+    // Sanity check just in case.
+    if (val.x != 19) {
+      throw new IllegalStateException("something went wrong during larval construction");
+    }
+  }
+
+  private static void stressTheGC() {
```

Comment:
There's a WhiteBox FullGC that you can call that might be a reliable way of calling System.gc().

Reply:
Thanks for letting me know, I've pushed an update.